### PR TITLE
created hand-gesture-detection.py

### PR DESCRIPTION
**Pull request by aishwarya24bhagwat**
added hand-gesture-detection for computer vision project, it detects movement of the hand and stop action, hand open and close (both right and left hand), clockwise and counter-clockwise direction of finger

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: computer_vision/hand-gesture-detection.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+import cv2 as cv
+import numpy as np
+import mediapipe as mp
+
+from utils import CvFpsCalc
+from model import KeyPointClassifier
+from model import PointHistoryClassifier
+
+# get_args() -> argparse.Namespace:
+def get_args():  #This is an utility function for parsing arguments
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:19]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+                        default=0.5)
+
+    args = analyzer.parse_args()
+    
+   
+    return args
+    
+
+# main() -> None:
+def main():
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:43]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+    with open(
+            'model/point_history_classifier/point_history_classifier_label.csv',
+            encoding='utf-8-sig') as f:
+        point_history_classifier_labels = csv.reader(f)
+        point_history_classifier_labels = [
+            row[0] for row in point_history_classifier_labels
+        ]
+
+    #measure frames per second-fps
+    cvFpsCalc = CvFpsCalc(buffer_len=10)
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:91]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+        debug_image = draw_point_history(debug_image, point_history)
+        debug_image = draw_info(debug_image, fps, mode, number)
+
+        cv.imshow('Hand Gesture Recognition', debug_image) #screening of reflection
+
+    cap.release()
+    cv.destroyAllWindows()
+
+
+def select_mode(key, mode):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:185]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        debug_image = draw_info(debug_image, fps, mode, number)
+
+        cv.imshow('Hand Gesture Recognition', debug_image) #screening of reflection
+
+    cap.release()
+    cv.destroyAllWindows()
+
+# key - key of select mode: 
+# mode - mode of select mode:
+def select_mode(key, mode):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:185]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        mode = 0
+    if key == 107:  # k
+        mode = 1
+    if key == 104:  # h
+        mode = 2
+    return number, mode
+
+# image - shape of image
+# landmark - array of landmark 
+def calc_bounding_rect(image, landmarks):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:199]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+        landmark_point = [np.array((landmark_x, landmark_y))]
+
+        landmark_array = np.append(landmark_array, landmark_point, axis=0)
+
+    x, y, w, h = cv.boundingRect(landmark_array)
+
+    return [x, y, x + w, y + h]
+
+
+def calc_landmark_list(image, landmarks):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:217]: added comment as requested

[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:217]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+        landmark_x = min(int(landmark.x * image_width), image_width - 1)
+        landmark_y = min(int(landmark.y * image_height), image_height - 1)
+        # landmark_z = landmark.z
+
+        landmark_point.append([landmark_x, landmark_y])
+
+    return landmark_point
+
+
+def pre_process_landmark(landmark_list):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:234]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+        temp_landmark_list[index][1] = temp_landmark_list[index][1] - base_y
+
+    # Convert to a one-dimensional list
+    temp_landmark_list = list(
+        itertools.chain.from_iterable(temp_landmark_list))
+
+    # Normalization
+    max_value = max(list(map(abs, temp_landmark_list)))
+
+    def normalize_(n):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:253]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+        return n / max_value
+
+    temp_landmark_list = list(map(normalize_, temp_landmark_list))
+
+    return temp_landmark_list
+
+
+def pre_process_point_history(image, point_history):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:261]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+                                        base_y) / image_height
+
+    # Convert to a one-dimensional list
+    temp_point_history = list(
+        itertools.chain.from_iterable(temp_point_history))
+
+    return temp_point_history
+
+
+def logging_csv(number, mode, landmark_list, point_history_list):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:284]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 12]
```diff
@@ -0,0 +1,541 @@
...
+            writer.writerow([number, *landmark_list])
+    if mode == 2 and (0 <= number <= 9):
+        csv_path = 'model/point_history_classifier/point_history.csv'
+        with open(csv_path, 'a', newline="") as f:
+            writer = csv.writer(f)
+            writer.writerow([number, *point_history_list])
+    return
+
+
+def draw_landmarks(image, landmark_point):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:300]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+            cv.circle(image, (landmark[0], landmark[1]), 5, (0, 0, 0), 1)
+        if index == 20:  
+            cv.circle(image, (landmark[0], landmark[1]), 8, (255, 255, 255),
+                      -1)
+            cv.circle(image, (landmark[0], landmark[1]), 8, (0, 0, 0), 1)
+
+    return image
+
+
+def draw_bounding_rect(use_brect, image, brect):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:488]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+            cv.circle(image, (landmark[0], landmark[1]), 5, (0, 0, 0), 1)
+        if index == 20:  
+            cv.circle(image, (landmark[0], landmark[1]), 8, (255, 255, 255),
+                      -1)
+            cv.circle(image, (landmark[0], landmark[1]), 8, (0, 0, 0), 1)
+
+    return image
+
+# draw_bounding_rect() -> jpg:
+def draw_bounding_rect(use_brect, image, brect):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:488]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+    if use_brect:
+        # Outer rectangle
+        cv.rectangle(image, (brect[0], brect[1]), (brect[2], brect[3]),
+                     (0, 0, 0), 1)
+
+    return image
+
+# draw_info_text() -> jpg:
+def draw_info_text(image, brect, handedness, hand_sign_text,
+                   finger_gesture_text):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:498]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+        cv.putText(image, "Finger Gesture:" + finger_gesture_text, (10, 60),
+                   cv.FONT_HERSHEY_SIMPLEX, 1.0, (0, 0, 0), 4, cv.LINE_AA)
+        cv.putText(image, "Finger Gesture:" + finger_gesture_text, (10, 60),
+                   cv.FONT_HERSHEY_SIMPLEX, 1.0, (255, 255, 255), 2,
+                   cv.LINE_AA)
+
+    return image
+
+
+def draw_point_history(image, point_history):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:518]: added comment as requested

File: computer_vision/hand-gesture-detection.py [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+    for index, point in enumerate(point_history):
+        if point[0] != 0 and point[1] != 0:
+            cv.circle(image, (point[0], point[1]), 1 + int(index / 2),
+                      (152, 251, 152), 2)
+
+    return image
+
+
+def draw_info(image, fps, mode, number):
```
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:527]: added comment as requested

File: computer_vision/hand-gesture-detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aishwarya24bhagwat) [reply computer_vision/hand-gesture-detection.py:19]: added comment as requested

**Review by AnkushDaharwal07 — APPROVED** [review]

**Comment by cclauss** [comment]
In preparation for that denial-of-service attack that is [Hacktoberfest](https://hacktoberfest.com), I am closing all pull requests that have failing tests.

**Pull request closed without merging**